src/GafferUI/FPSGadget.cpp

```diff
@@ -0,0 +1,115 @@
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferUI/FPSGadget.h"
+
+#include "GafferUI/Style.h"
+#include "GafferUI/ViewportGadget.h"
+
+using namespace GafferUI;
+
+GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( FPSGadget );
+
+FPSGadget::FPSGadget()
+	:	Gadget( defaultName<FPSGadget>() ), m_totalTime( 0 )
+{
+}
+
+FPSGadget::~FPSGadget()
+{
+}
+
+void FPSGadget::renderLayer( Gadget::Layer layer, const Style *style, Gadget::RenderReason reason ) const
+{
+	if( layer != Layer::Main )
+	{
+		return;
+	}
+
+	if( isSelectionRender( reason ) )
+	{
+		return;
+	}
+
+	static std::chrono::time_point<std::chrono::high_resolution_clock> lastTime;
+	static bool initialized = false;
+
+	auto curTime = std::chrono::high_resolution_clock::now();
+	int frameTime = std::chrono::duration_cast< std::chrono::microseconds >( curTime - lastTime ).count();
+	lastTime = curTime;
+
+	if( !initialized )
+	{
+		initialized = true;
+		return;
+	}
+
+	// Simple running average
+	m_totalTime += frameTime;
+	m_timeBuffer.push_back( frameTime );
+
+	// 20 frames is enough to smooth out bumps, while not lagging behind too much
+	while( m_timeBuffer.size() > 20 )
+	{
+		m_totalTime -= m_timeBuffer.front();
+		m_timeBuffer.pop_front();
+	}
+
+	float averageFrameTime = m_totalTime / ( m_timeBuffer.size() * 1000000.0f );
+
+	const ViewportGadget *viewportGadget = ancestor<ViewportGadget>();
+	ViewportGadget::RasterScope rasterScope( viewportGadget );
+
+	glTranslatef( 5, 50, 0.0f );
```

Comment:
This seems like a reasonable position for the Viewer, but less so for the GraphEditor (or other potential usage points). Should we accessors to allow it to be specified? Or maybe we should just use `getTransform()` so that `setTransform()` can be used to provide an arbitrary raster-space transform?

Reply:
Lets see, I don't see any need to change it dynamically, could probably just be a constructor argument, defaulting to this position?

Reply:
Hmm, that's fair.  My thought was just that I would prefer to have it show up somewhere reasonable if you don't override the transform.  Is it reasonable to have a constructor argument with a default value that calls setTransform?

Reply:
> My thought was just that I would prefer to have it show up somewhere reasonable if you don't override the transform.
 
 Fair point - I guess the identity transform would put it in the top left, but rendering just outside of frame?
 
 > Is it reasonable to have a constructor argument with a default value that calls setTransform?

The whole subject of raster space gadgets is a bit of an odd one really. I think maybe a more principled approach would be to move the RasterScope out of the Gadgets themselves, and have the ViewportGadget manage separate containers of raster and non-raster gadgets, organising the appropriate transform scope for each before drawing. Then FPSGadget could report a proper bound, and we'd just put it in a standard container gadget that would be responsible for performing the layout. That's a whole other topic though...

Passing a transform to the constructor would definitely be an oddity as far as other gadgets go, but it seems like a reasonable compromise here...